### ISSUE TITLE
It should be possible to calculate dissimilarity on a user-defined list of features

### ISSUE THREAD

**Issue by marijnkeijzer** [marijnkeijzer/defSim#31]
Dissimilarity between agents is currently calculated based on all agent features. However, more complex attribute initializers might demand that agents have characteristics that govern how their attributes are initialized, but do not matter when calculating (opinion-)dissimilarity. 

The challenge is that the dissimilarity calculator is called in initialization ánd at each iteration by the influence_function. This means that the list of features on which dissimilarity should be calculated needs to be passed to the spread_influence function. This means a major and uncertain revision of Experiment, Simulation, the dissimilarity component, and the influence component. Or is there another way?

**Comment by marijnkeijzer** [comment]
I'm thinking that maybe it would be smart to define a list of features that should be excluded by the dissimilarity calculator rather than included. This makes handling the default case (where the user does not define this) easier.

**Comment by dimaba** [comment]
I think there are two ways we can do this. We can add keyword arguments for this, which would already get passed through Experiment, to Simulation, to spread_influence, to the influence function. Each dissimilarity calculator can then look for these keyword arguments. We would then need to make a change in the `__init__` method of the Simulation to also pass these keyword arguments along for the network initialization. That should work.

However, we're at a point where we should probably start limiting the number of parameters users can set for each simulation. It's not technically much of a problem to pass along this long list of parameters throughout the simulation, but it isn't elegant and we have to start worrying about conflicting names for parameters that end up being used in different components. (For example, I'm also considering that the new parameter keywords I added in the random continuous initializer are already too much.)

Fortunately, we already allow users to pass an instance of a DissimilarityCalculator here. So an alternative would be to just create attributes on each currently implemented DissimilarityCalculator. We could even have two: one for features to include, and one for features to exclude. In the calculate_dissimilarity function, we check whether these attributes are not None. If neither is set, we default to using all features. Then a user just has to create an instance of, for example, EuclideanDistance with these attributes set and use that as dissimilarity calculator.

One downside is that it currently would not be possible to set a parameter that varies which features count for dissimilarity in an experiment. (We don't accept lists of DissimilarityCalculator for example.) To solve that, there is always the solution that a user creates simulations with different dissimilarity calculators and then uses that list of simulations as basis for the experiment.

More generally, it might be a good idea to talk about how we deal with the customization of different elements. What becomes a parameter you can set in parameter_dict, and what doesn't. One line we could draw is that the parameters which are relevant to the models of influence we implement in the defSim core go in parameter_dict, and customization beyond that goes in the modules (instances with attributes, or custom classes). But that's just a thought.

**Comment by marijnkeijzer** [comment]
Thanks. That was very helpful. I couldn't agree more about restricting the number of kwargs that we pass around all the time. However, if we want the user to write their own dissimilarity calculator, this component does need to receive the kwargs dictionary. That was not yet possible. 

In a new branch I now included the revision. It works with the code example below (which I wrote for the usecase examples, which triggered this revision in the first place). It does not yet work with the HammingDistance calculator, but the main idea is there.

```python
import defSim as ds
from defSim.agents_init.agents_init import AttributesInitializer
from defSim.agents_init.agents_init import set_categorical_attribute
import numpy as np

class MyAttributeInitializer(AttributesInitializer):
    @staticmethod
    def initialize_attributes(network, **kwargs):
        set_categorical_attribute(network, 'sex', ["female", "male"])
        for j in range(kwargs.get('num_features', 1)):
            for i in network.nodes():
                if network.nodes[i]['sex'] == "female":
                    network.nodes[i]['f{:02d}'.format(j+1)] = np.random.beta(1.5, 5)
                else:
                    network.nodes[i]['f{:02d}'.format(j+1)] = np.random.beta(5, 1.5)

my_experiment = ds.Experiment(attributes_initializer=MyAttributeInitializer,
                              attribute_parameters={'num_features':3},
                              dissimilarity_parameters={'dissimilarity_exclude':['sex']},
                              dissimilarity_measure="euclidean",
                              topology="complete_graph",
                              network_parameters={'n':20},
                              influenceable_attributes=['f01','f02','f03'],
                              influence_function="bounded_confidence",
                              influence_parameters={'confidence_level':[.3,.7]},
                              repetitions=3,
                              max_iterations=1000)
my_results = my_experiment.run()

print(my_results)
```

**Comment by dimaba** [comment]
Okay! Looks like that works. 
My alternative to avoid parameter dict would be something like:
```python
class EuclideanDistance(DissimilarityCalculator):
    # documentation omitted 

    def __init__(self, exclude = None):
    # documentation omitted
        self.exclude = exclude

    @staticmethod
    def calculate_dissimilarity(network: nx.Graph, agent1_id: int, agent2_id: int,
                                **kwargs) -> float:
    # documentation omitted

        agent1_attributes = [v for k, v in network.nodes[agent1_id].items() if k not in self.exclude]
        agent2_attributes = [v for k, v in network.nodes[agent2_id].items() if k not in self.exclude]
    
        # rest is the same
```

And then 
```python
my_experiment = ds.Experiment(attributes_initializer=MyAttributeInitializer,
                              attribute_parameters={'num_features':3},
                              dissimilarity_measure=EuclideanDistance(exclude = ['sex']),
                              #etc
```

Which I like syntax-wise because it makes the link between the parameter and the dissimilarity measure obvious and avoids the parameter dict, but has its own downsides. Mainly, currently, that varying parameters in an experiment can only be done through the parameter dictionaries which this does not use.

**Comment by marijnkeijzer** [comment]
Thanks. I like the solution without extra parameter dictionary. It is not likely that a parameter passed to this component needs to be varied in an experiment so I agree it's way cleaner that way. 

I can't seem to make the code above work though. PyCharm gives me the error "unresolved reference 'self' " in 'calculate_dissimilarity' and it's a bit hard for me to fix because I don't fully understand the logic behind referencing self here.

**Comment by marijnkeijzer** [comment]
Do you have any clue why?

**Comment by marijnkeijzer** [comment]
Solved 2b187547db9c10f263417dfb7ed64d44142c1f26
